File: algorithms/0897.increasing-order-search-tree/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}
def increasing_bst(root)
  nodes = build_node_array(root)

  for i in 0...nodes.size
    nodes[i].left = nil
    nodes[i].right = nodes[i + 1]
  end

  nodes[0]
end

def build_node_array(root)
  return [] unless root
  build_node_array(root.left) + [root] + build_node_array(root.right)
end

File: algorithms/0897.increasing-order-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func increasingBST(root *TreeNode) *TreeNode {
	values := []int{}
	helper(root, &values)

	res := &TreeNode{Val: values[0], Left: nil, Right: nil}
	tmp := res
	for i := 1; i < len(values); i++ {
		tmp.Right = &TreeNode{Val: values[i], Left: nil, Right: nil}
		tmp = tmp.Right
	}

	return res
}

func helper(root *TreeNode, values *[]int) {
	if root == nil {
		return
	}

	helper(root.Left, values)
	*values = append(*values, root.Val)
	helper(root.Right, values)
}

